tools/ckb4ibc-test/src/framework/utils/ckb.rs

```diff
@@ -0,0 +1,412 @@
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_fmt(format_args!("{}", self.result))
+    }
+}
+
+pub fn check_channel(channel: &IbcChannel) -> bool {
+    channel.state == State::Open
+}
+
+pub fn check_ibc_connection(connection: IbcConnections, count: u32) -> bool {
+    if connection.connections.len() != count as usize {
+        return false;
+    }
+    // FIXME @jjy in current implementation,
+    // a failed init conn is remained in the vector, so we must check last opened conn
+    let connection = connection.connections.into_iter().last().unwrap();
+    connection.state == State::Open
+}
+
+fn send_tx(req: &str, port: u32) -> RpcResponse {
+    let output = Command::new("curl")
+        .arg("-H")
+        .arg("content-type: application/json")
+        .arg("-d")
+        .arg(format!("@{}", req))
+        .arg(format!("http://localhost:{}", port))
+        .output()
+        .unwrap();
+    serde_json::from_slice(&output.stdout)
+        .map_err(|e| {
+            println!("{}", String::from_utf8_lossy(&output.stdout));
+            e
+        })
+        .unwrap()
+}
+
+pub fn prepare_ckb_chain(
+    ckb_path: &str,
+    port: u32,
+    genesis_wallets: &[(&Wallet, u64)],
+) -> (ChildProcess, ChildProcess) {
+    println!("\n========== Prepare Ckb node on port {port} ==========\n");
+
+    let mut working_dir = std::env::current_dir().unwrap();
+    working_dir.push(ckb_path);
+
+    let _ = std::fs::remove_dir_all(ckb_path);
+    std::fs::create_dir(ckb_path).unwrap();
+
+    Command::new("ckb")
+        .arg("init")
+        .arg("--chain")
+        .arg("dev")
+        .current_dir(&working_dir)
+        .spawn()
+        .unwrap()
+        .wait()
+        .unwrap();
+
+    std::fs::copy(
+        "../forcerelay-test/ckb/ckb.toml",
+        format!("{}/ckb.toml", ckb_path),
+    )
+    .unwrap();
+    std::fs::copy(
+        "../forcerelay-test/ckb/dev.toml",
+        format!("{}/specs/dev.toml", ckb_path),
+    )
+    .unwrap();
+    std::fs::copy("ckb-miner.toml", format!("{}/ckb-miner.toml", ckb_path)).unwrap();
+
+    // issue cells
+    {
+        let dev_spec_path = format!("{}/specs/dev.toml", ckb_path);
+        let content = fs::read_to_string(&dev_spec_path).unwrap();
+        let mut dev_spec: ChainSpec = toml::from_str(&content).expect("invalid spec");
+        let issued_cell_template = dev_spec.genesis.issued_cells.last().unwrap().to_owned();
+
+        for (wallet, _capacity) in genesis_wallets.iter() {
+            let args = pubkey_to_script_args(&wallet.key.public_key);
+            let mut cell = issued_cell_template.clone();
+            cell.lock.args = JsonBytes::from_vec(args.to_vec());
+            dev_spec.genesis.issued_cells.push(cell);
+        }
+        let content = toml::to_string_pretty(&dev_spec).unwrap();
+        fs::write(dev_spec_path, content).unwrap();
+    }
+
+    if port != 8114 {
+        Command::new("sed")
```

Comment:
Is it possible to use `toml_edit` instead?

Reply:
Agreed, but this is also copied from the old code, I want to fix it in later PRs.